src/Microsoft.AspNet.OData.Shared/Query/ExpandQueryBuilder.cs

```diff
@@ -0,0 +1,110 @@
+﻿//-----------------------------------------------------------------------------
+// <copyright file="ExpandQueryBuilder.cs" company=".NET Foundation">
+//      Copyright (c) .NET Foundation and Contributors. All rights reserved. 
+//      See License.txt in the project root for license information.
+// </copyright>
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.AspNet.OData.Formatter;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Exposes the ability to generate a $expand query string from a payload object.
+    /// </summary>
+    public class ExpandQueryBuilder : IExpandQueryBuilder
+    {
+        /// <inheritdoc />
+        public virtual string GenerateExpandQueryString(object value, IEdmModel model)
+        {
+            return GenerateExpandQueryStringInternal(value, model, true);
+        }
+
+        private string GenerateExpandQueryStringInternal(object value, IEdmModel model, bool isFirstExpand)
+        {
+            Type type = value.GetType();
+            Type _elementType = null;
+            bool isCollection = TypeHelper.IsCollection(type, out _elementType);
+            IList objList = null;
+            if (isCollection)
+            {
+                type = _elementType;
+                objList = value as IList;
```

Comment:
this can result to a null? do we have a check for it?

Reply:
Got it.
Fixed